UI/source-tree.cpp

```diff
@@ -1075,6 +1096,119 @@ void SourceTree::UpdateIcons()
 	stm->SceneChanged();
 }
 
+void SourceTree::dragEnterEvent(QDragEnterEvent *event)
+{
+	if (event->mimeData()->hasFormat("application/indexes")) {
+		event->acceptProposedAction();
+	} else {
+		event->ignore();
+	}
+}
+
+void SourceTree::dragMoveEvent(QDragMoveEvent *event)
+{
+	if (event->mimeData()->hasFormat("application/indexes")) {
+		event->acceptProposedAction();
+	} else {
+		event->ignore();
+	}
+}
+
+void SourceTree::mousePressEvent(QMouseEvent *event)
+{
+	if (event->button() == Qt::LeftButton) {
+		dragStartPosition = event->pos();
+	}
+	QListView::mousePressEvent(event);
+}
+
+void SourceTree::startDrag(Qt::DropActions supportedActions)
+{
+	obs_source_t *current_scene_source = obs_frontend_get_current_scene();
+	if (!current_scene_source) {
+		return;
+	}
+
+	obs_scene_t *current_scene =
+		obs_scene_from_source(current_scene_source);
+	if (!current_scene) {
+		obs_source_release(current_scene_source);
+		return;
+	}
+
+	QStringList selectedSourceNames;
+	obs_scene_enum_items(current_scene, get_selected_source_names,
+			     &selectedSourceNames);
+	obs_source_release(current_scene_source);
+
+	if (selectedSourceNames.isEmpty()) {
+		return;
+	}
+
+	QByteArray itemData;
+	QDataStream dataStream(&itemData, QIODevice::WriteOnly);
+	dataStream << selectedSourceNames;
+	QMimeData *mimeData = new QMimeData;
+	mimeData->setData("application/indexes", itemData);
+
+	QDrag *drag = new QDrag(this);
+	drag->setMimeData(mimeData);
+
+	QRect boundingRect;
+	QModelIndexList indexes = selectedIndexes();
+	for (const QModelIndex &index : indexes) {
+		QRect rect = visualRect(index);
+		boundingRect = boundingRect.isNull()
+				       ? rect
+				       : boundingRect.united(rect);
+	}
+
+	QPixmap pixmap(boundingRect.size());
+	pixmap.fill(Qt::transparent);
+	QPainter painter(&pixmap);
+	painter.setRenderHint(QPainter::Antialiasing, true);
+
+	for (const QModelIndex &index : indexes) {
+		QRect rect = visualRect(index);
+		painter.save();
+		painter.translate(rect.topLeft() - boundingRect.topLeft());
+		itemDelegate()->paint(&painter, QStyleOptionViewItem(), index);
+		painter.restore();
+	}
+	painter.end();
+
+	QPixmap blueBox(pixmap.size());
+	blueBox.fill(QColor(0, 0, 153, 127));
+	QPainter boxPainter(&blueBox);
+	boxPainter.drawPixmap(0, 0, pixmap);
+	boxPainter.end();
+
+	drag->setPixmap(blueBox);
+	drag->setHotSpot(dragStartPosition - boundingRect.topLeft());
+
+	Qt::DropAction dropAction = drag->exec(supportedActions);
+}
+
+
+bool SourceTree::get_selected_source_names(obs_scene_t *scene,
```

Comment:
```suggestion
static get_selected_source_names(obs_scene_t */*scene*/,
```

This static function has no reason to be made public.